drivers/vdpa/virtio/virtio_vdpa.c

```diff
@@ -0,0 +1,955 @@
+	.reset_stats = NULL,
+};
+
+static int vdpa_check_handler(__rte_unused const char *key,
+		const char *value, void *ret_val)
+{
+	if (strcmp(value, "1") == 0)
+		*(int *)ret_val = 1;
+	else
+		*(int *)ret_val = 0;
+
+	return 0;
+}
+
+#define VIRTIO_ARG_VDPA 	  "vdpa"
+
+static int
+virtio_pci_devargs_parse(struct rte_devargs *devargs, int *vdpa)
+{
+	struct rte_kvargs *kvlist;
+	int ret = 0;
+
+	if (devargs == NULL)
+		return 0;
+
+	kvlist = rte_kvargs_parse(devargs->args, NULL);
+	if (kvlist == NULL) {
+		DRV_LOG(ERR, "Error when parsing param");
+		return 0;
+	}
+
+	if (rte_kvargs_count(kvlist, VIRTIO_ARG_VDPA) == 1) {
+		/* vdpa mode selected when there's a key-value pair:
+		 * vdpa=1
+		 */
+		ret = rte_kvargs_process(kvlist, VIRTIO_ARG_VDPA,
+				vdpa_check_handler, vdpa);
+		if (ret < 0)
+			DRV_LOG(ERR, "Failed to parse %s", VIRTIO_ARG_VDPA);
+	}
+
+	rte_kvargs_free(kvlist);
+
+	return ret;
+}
+
+static void
+virtio_vdpa_queues_free(struct virtio_vdpa_priv *priv)
+{
+	uint16_t nr_vq = priv->hw_nr_virtqs;
+	struct virtio_vdpa_vring_info *vr;
+	uint16_t i;
+
+	if(priv->vrings) {
+		for (i = 0; i < nr_vq; i++) {
+			vr = priv->vrings[i];
+			if (!vr)
+				continue;
+			rte_free(vr);
+			priv->vrings[i] = NULL;
+		}
+		rte_free(priv->vrings);
+		priv->vrings = NULL;
+	}
+
+	virtio_pci_dev_queues_free(priv->vpdev,nr_vq);
+
+}
+
+static int
+virtio_vdpa_queues_alloc(struct virtio_vdpa_priv *priv)
+{
+	uint16_t nr_vq = priv->hw_nr_virtqs;
+	struct virtio_vdpa_vring_info *vr;
+	uint16_t i;
+	int ret;
+
+	ret = virtio_pci_dev_queues_alloc(priv->vpdev, nr_vq);
+	if (ret) {
+		DRV_LOG(ERR, "Failed to alloc virtio device queues.");
+		return ret;
+	}
+
+	priv->vrings = rte_zmalloc(NULL, sizeof(struct virtio_vdpa_vring_info *) * nr_vq, 0);
+	if (!priv->vrings) {
+		virtio_vdpa_queues_free(priv);
+		return -ENOMEM;
+	}
+
+	for (i = 0; i < nr_vq; i++) {
+		vr = rte_zmalloc_socket(NULL, sizeof(struct virtio_vdpa_vring_info), RTE_CACHE_LINE_SIZE,
+					priv->pdev->device.numa_node);
+		if (vr == NULL) {
+			virtio_vdpa_queues_free(priv);
+			return -ENOMEM;
+		}
+		priv->vrings[i] = vr;
+		priv->vrings[i]->index = i;
+		priv->vrings[i]->priv = priv;
+	}
+	return 0;
+}
+
+static int
+virtio_vdpa_dev_probe(struct rte_pci_driver *pci_drv __rte_unused,
+		struct rte_pci_device *pci_dev)
+{
+	int vdpa = 0;
+	int ret = 0;
+	struct virtio_vdpa_priv *priv = NULL;
+	char devname[RTE_DEV_NAME_MAX_LEN] = {0};
+	int iommu_group_num;
+
+	ret = virtio_pci_devargs_parse(pci_dev->device.devargs, &vdpa);
+	if (ret < 0) {
+		DRV_LOG(ERR, "Devargs parsing is failed");
+		return ret;
+	}
+	/* virtio vdpa pmd skips probe if device needs to work in none vdpa mode */
+	if (vdpa != 1)
+		return 1;
+
+	priv = rte_zmalloc("virtio vdpa device private", sizeof(*priv), RTE_CACHE_LINE_SIZE);
+	if (!priv) {
+		DRV_LOG(ERR, "Failed to allocate private memory.");
+		rte_errno = ENOMEM;
+		return -rte_errno;
+	}
+
+	//TO_DO: need to confirm following:
+	priv->vfio_dev_fd = -1;
```

Comment:
I think all of these should stay inside the virtio driver operating over vfio.
we at this driver layer only use it.

Reply:
DMA map operation still needs this